File: src/phantom/registry/scanner.py
"""
Project Scanner

Scans ~/arch directory to discover and register all projects.
Detects project types, languages, and collects initial metadata.
"""

import os
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional, Set
import json
import logging

from phantom.intelligence.core import (
    CerebroIntelligence,
    Project,
    ProjectStatus,
)
from phantom.intelligence.collectors import (
    SignalCollector,
    HumanIntelCollector,
    OpenSourceCollector,
    TechIntelCollector,
)

logger = logging.getLogger("cerebro.scanner")


# Directories to ignore during scanning
IGNORED_DIRS = {
    ".git",
    ".github",
    ".gitlab",
    "__pycache__",
    "node_modules",
    ".venv",
    "venv",
    "target",
    "dist",
    "build",
    ".cache",
    ".nix-profile",
    "result",
}

# Files that indicate a project root
PROJECT_MARKERS = [
    "flake.nix",        # Nix flake
    "Cargo.toml",       # Rust
    "package.json",     # Node.js
    "pyproject.toml",   # Python
    "go.mod",           # Go
    "pom.xml",          # Java Maven
    "build.gradle",     # Java Gradle
    "Makefile",         # Generic
    "justfile",         # Just task runner
    "README.md",        # Documentation
    ".git",             # Git repo
]


class ProjectScanner:
    """
    Scans and discovers projects in the ecosystem.

    Automatically detects:
    - Project boundaries
    - Languages used
    - Build systems
    - Configuration files
    """

    def __init__(
        self,
        cerebro: CerebroIntelligence,
        arch_path: str = "/home/kernelcore/arch",
    ):
        self.cerebro = cerebro
        self.arch_path = Path(arch_path)

        # Initialize collectors
        self.collectors = [
            SignalCollector(cerebro),
            HumanIntelCollector(cerebro),
            OpenSourceCollector(cerebro),
            TechIntelCollector(cerebro),
        ]

    def scan(self, full_scan: bool = False) -> List[Project]:
        """
        Scan the arch directory for projects.

        Args:
            full_scan: If True, re-scan all projects. If False, only scan new ones.

        Returns:
            List of discovered projects.
        """
        logger.info(f"Starting scan of {self.arch_path}")
        discovered_projects = []

        if not self.arch_path.exists():
            logger.warning(f"Arch path does not exist: {self.arch_path}")
            return discovered_projects

        # Scan top-level directories
        for item in self.arch_path.iterdir():
            if not item.is_dir():
                continue

            if item.name in IGNORED_DIRS or item.name.startswith("."):
                continue

            # Check if it's a project
            if self._is_project(item):
                project = self._scan_project(item)
                if project:
                    discovered_projects.append(project)
                    self.cerebro.register_project(project)
                    logger.info(f"Discovered project: {project.name}")

        logger.info(f"Scan complete. Found {len(discovered_projects)} projects.")
        return discovered_projects

    def _is_project(self, path: Path) -> bool:
        """Check if a directory is a project root."""
        for marker in PROJECT_MARKERS:
            if (path / marker).exists():
                return True
        return False

    def _scan_project(self, path: Path) -> Optional[Project]:
        """Scan a single project and extract metadata."""
        try:
            project = Project(
                name=path.name,
                path=path,
            )

            # Detect languages
            project.languages = self._detect_languages(path)

            # Get description from README
            project.description = self._extract_description(path)

            # Get last commit time
            project.last_commit = self._get_last_commit_time(path)

            # Calculate initial health score
            project.health_score = self._calculate_initial_health(path)

            # Determine status
            project.status = self._determine_status(project)

            # Set indexed time
            project.last_indexed = datetime.now(timezone.utc)

            # Collect metadata
            project.metadata = self._collect_metadata(path)

            return project

        except Exception as e:
            logger.error(f"Error scanning project {path.name}: {e}")
            return None

    def _detect_languages(self, path: Path) -> List[str]:
        """Detect programming languages used in the project."""
        languages = set()

        # Check for language-specific files
        language_indicators = {
            "*.py": "Python",
            "*.rs": "Rust",
            "*.ts": "TypeScript",
            "*.tsx": "TypeScript",
            "*.js": "JavaScript",
            "*.jsx": "JavaScript",
            "*.go": "Go",
            "*.nix": "Nix",
            "*.sol": "Solidity",
            "*.java": "Java",
            "*.cpp": "C++",
            "*.c": "C",
            "*.zig": "Zig",
            "*.svelte": "Svelte",
            "*.vue": "Vue",
        }

        for pattern, language in language_indicators.items():
            files = list(path.rglob(pattern))
            # Filter out ignored directories
            files = [f for f in files if not any(
                ignored in f.parts for ignored in IGNORED_DIRS
            )]
            if files:
                languages.add(language)

        # Check for config files
        if (path / "Cargo.toml").exists():
            languages.add("Rust")
        if (path / "package.json").exists():
            languages.add("JavaScript")
        if (path / "pyproject.toml").exists() or (path / "setup.py").exists():
            languages.add("Python")
        if (path / "go.mod").exists():
            languages.add("Go")
        if (path / "flake.nix").exists():
            languages.add("Nix")

        return list(languages)

    def _extract_description(self, path: Path) -> str:
        """Extract project description from README."""
        readme_files = ["README.md", "README.rst", "README.txt", "README"]

        for readme_name in readme_files:
            readme_path = path / readme_name
            if readme_path.exists():
                try:
                    content = readme_path.read_text()
                    # Extract first paragraph after title
                    lines = content.split("\n")
                    description_lines = []
                    in_description = False

                    for line in lines:
                        # Skip title lines
                        if line.startswith("#"):
                            in_description = True
                            continue
                        # Skip empty lines at start
                        if in_description and line.strip():
                            description_lines.append(line.strip())
                            if len(" ".join(description_lines)) > 200:
                                break

                    return " ".join(description_lines)[:300]
                except Exception:
                    pass

        return ""

    def _get_last_commit_time(self, path: Path) -> Optional[datetime]:
        """Get the timestamp of the last git commit."""
        if not (path / ".git").exists():
            return None

        try:
            result = subprocess.run(
                ["git", "log", "-1", "--format=%ct"],
                cwd=path,
                capture_output=True,
                text=True,
                timeout=10,
            )

            if result.returncode == 0 and result.stdout.strip():
                timestamp = int(result.stdout.strip())
                return datetime.fromtimestamp(timestamp, tz=timezone.utc)
        except Exception:
            pass

        return None

    def _calculate_initial_health(self, path: Path) -> float:
        """Calculate initial health score for a project."""
        score = 50.0  # Start at 50%

        # Has README? +10
        if any((path / f).exists() for f in ["README.md", "README.rst", "README"]):
            score += 10

        # Has tests? +15
        if (path / "tests").exists() or (path / "test").exists():
            score += 15

        # Has CI? +10
        if (path / ".github" / "workflows").exists() or (path / ".gitlab-ci.yml").exists():
            score += 10

        # Has flake.nix? +5 (reproducible)
        if (path / "flake.nix").exists():
            score += 5

        # Has docs? +10
        if (path / "docs").exists():
            score += 10

        return min(100.0, score)

    def _determine_status(self, project: Project) -> ProjectStatus:
        """Determine project status based on activity."""
        if not project.last_commit:
            return ProjectStatus.UNKNOWN

        days_since_commit = (
            datetime.now(timezone.utc) - project.last_commit
        ).days

        if days_since_commit < 30:
            return ProjectStatus.ACTIVE
        elif days_since_commit < 90:
            return ProjectStatus.MAINTENANCE
        elif days_since_commit < 365:
            return ProjectStatus.DEPRECATED
        else:
            return ProjectStatus.ARCHIVED

    def _collect_metadata(self, path: Path) -> Dict[str, Any]:
        """Collect additional metadata about the project."""
        metadata = {}

        # Check for Nix flake
        flake_path = path / "flake.nix"
        if flake_path.exists():
            metadata["has_flake"] = True

        # Check for Docker
        if (path / "Dockerfile").exists() or (path / "docker-compose.yml").exists():
            metadata["has_docker"] = True

        # Check for Justfile
        if (path / "justfile").exists() or (path / "Justfile").exists():
            metadata["has_justfile"] = True

        # Count files
        try:
            file_count = sum(1 for _ in path.rglob("*") if _.is_file())
            metadata["file_count"] = file_count
        except Exception:
            metadata["file_count"] = 0

        return metadata

    def collect_intelligence(self, project: Project) -> int:
        """Run all collectors on a project."""
        total_items = 0

        for collector in self.collectors:
            try:
                items = collector.collect(project)
                for item in items:
                    self.cerebro.add_intelligence(item)
                    total_items += 1
            except Exception as e:
                logger.error(f"Collector {collector.__class__.__name__} failed: {e}")

        logger.info(f"Collected {total_items} intelligence items from {project.name}")
        return total_items

    def full_scan_with_intelligence(self) -> Dict[str, Any]:
        """
        Perform a full scan with intelligence collection.

        Returns scan statistics.
        """
        stats = {
            "projects_found": 0,
            "intelligence_collected": 0,
            "errors": 0,
            "duration_seconds": 0,
        }

        start_time = datetime.now(timezone.utc)

        # Scan for projects
        projects = self.scan(full_scan=True)
        stats["projects_found"] = len(projects)

        # Collect intelligence from each project
        for project in projects:
            try:
                items = self.collect_intelligence(project)
                stats["intelligence_collected"] += items
            except Exception as e:
                logger.error(f"Failed to collect intel from {project.name}: {e}")
                stats["errors"] += 1

        # Save state
        self.cerebro.save()

        end_time = datetime.now(timezone.utc)
        stats["duration_seconds"] = (end_time - start_time).total_seconds()

        logger.info(
            f"Full scan complete: {stats['projects_found']} projects, "
            f"{stats['intelligence_collected']} intel items, "
            f"{stats['errors']} errors, "
            f"{stats['duration_seconds']:.1f}s"
        )

        return stats

File: crates/api-server/src/dashboard/scanner.rs
use anyhow::{Context, Result};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use tokio::fs;
use tracing::{debug, info, warn};

use super::models::{Project, ProjectStatus};

pub struct ProjectScanner {
    base_path: PathBuf,
}

impl ProjectScanner {
    pub fn new(base_path: PathBuf) -> Self {
        Self { base_path }
    }

    /// Scan the base directory for projects
    pub async fn scan(&self) -> Result<Vec<Project>> {
        info!("📂 Scanning projects in {:?}", self.base_path);

        let mut projects = Vec::new();
        let mut entries = fs::read_dir(&self.base_path).await?;

        while let Some(entry) = entries.next_entry().await? {
            let path = entry.path();

            // Skip hidden directories
            if let Some(name) = path.file_name() {
                if name.to_string_lossy().starts_with('.') {
                    continue;
                }
            }

            if path.is_dir() {
                match self.analyze_directory(&path).await {
                    Ok(Some(project)) => {
                        debug!("✓ Found project: {}", project.name);
                        projects.push(project);
                    }
                    Ok(None) => {
                        debug!("⊘ Skipped: {:?} (not a project)", path);
                    }
                    Err(e) => {
                        warn!("⚠ Error analyzing {:?}: {}", path, e);
                    }
                }
            }
        }

        info!("✓ Found {} projects", projects.len());
        Ok(projects)
    }

    /// Analyze a directory to determine if it's a project
    async fn analyze_directory(&self, path: &Path) -> Result<Option<Project>> {
        let name = path
            .file_name()
            .context("No filename")?
            .to_string_lossy()
            .to_string();

        // Check if it's a project directory (has git, cargo, package.json, etc.)
        let is_project = self.is_project_directory(path).await?;
        if !is_project {
            return Ok(None);
        }

        // Detect languages
        let languages = self.detect_languages(path).await?;

        // Get description from README if available
        let description = self.extract_description(path).await?;

        // Determine project status
        let status = self.determine_status(path).await?;

        // Get last commit time
        let last_commit = self.get_last_commit(path).await?;

        let project = Project {
            name: name.clone(),
            path: path.to_string_lossy().to_string(),
            description,
            languages,
            status,
            health_score: 0.0, // Will be calculated by analyzer
            last_commit,
            last_indexed: Some(chrono::Utc::now().to_rfc3339()),
            metadata: HashMap::new(),
        };

        Ok(Some(project))
    }

    /// Check if directory is a project
    async fn is_project_directory(&self, path: &Path) -> Result<bool> {
        // Check for common project indicators
        let indicators = [
            ".git",
            "Cargo.toml",
            "package.json",
            "pyproject.toml",
            "go.mod",
            "pom.xml",
            "build.gradle",
        ];

        for indicator in indicators {
            if path.join(indicator).exists() {
                return Ok(true);
            }
        }

        Ok(false)
    }

    /// Detect programming languages in the project
    async fn detect_languages(&self, path: &Path) -> Result<Vec<String>> {
        let mut languages = Vec::new();

        // Check for language-specific files
        if path.join("Cargo.toml").exists() {
            languages.push("Rust".to_string());
        }
        if path.join("package.json").exists() {
            languages.push("JavaScript".to_string());
        }
        if path.join("tsconfig.json").exists() {
            languages.push("TypeScript".to_string());
        }
        if path.join("pyproject.toml").exists() || path.join("setup.py").exists() {
            languages.push("Python".to_string());
        }
        if path.join("go.mod").exists() {
            languages.push("Go".to_string());
        }
        if path.join("pom.xml").exists() {
            languages.push("Java".to_string());
        }

        // If no specific markers, try to detect from file extensions
        if languages.is_empty() {
            languages = self.detect_from_extensions(path).await?;
        }

        Ok(languages)
    }

    /// Detect languages from file extensions
    async fn detect_from_extensions(&self, path: &Path) -> Result<Vec<String>> {
        let mut language_counts: HashMap<String, usize> = HashMap::new();

        if let Ok(mut entries) = fs::read_dir(path).await {
            while let Ok(Some(entry)) = entries.next_entry().await {
                if let Some(ext) = entry.path().extension() {
                    let lang = match ext.to_str() {
                        Some("rs") => Some("Rust"),
                        Some("js") => Some("JavaScript"),
                        Some("ts") | Some("tsx") => Some("TypeScript"),
                        Some("py") => Some("Python"),
                        Some("go") => Some("Go"),
                        Some("java") => Some("Java"),
                        Some("c") | Some("h") => Some("C"),
                        Some("cpp") | Some("cc") | Some("cxx") => Some("C++"),
                        _ => None,
                    };

                    if let Some(lang) = lang {
                        *language_counts.entry(lang.to_string()).or_insert(0) += 1;
                    }
                }
            }
        }

        // Return languages sorted by count (most used first)
        let mut languages: Vec<_> = language_counts.into_iter().collect();
        languages.sort_by(|a, b| b.1.cmp(&a.1));
        Ok(languages.into_iter().map(|(lang, _)| lang).collect())
    }

    /// Extract description from README
    async fn extract_description(&self, path: &Path) -> Result<String> {
        let readme_paths = ["README.md", "README.txt", "README"];

        for readme_name in readme_paths {
            let readme_path = path.join(readme_name);
            if readme_path.exists() {
                if let Ok(content) = fs::read_to_string(&readme_path).await {
                    // Extract first meaningful line (not empty, not heading marker)
                    for line in content.lines() {
                        let line = line.trim();
                        if !line.is_empty() && !line.starts_with('#') {
                            return Ok(line.to_string());
                        }
                    }
                }
            }
        }

        Ok("No description available".to_string())
    }

    /// Determine project status
    async fn determine_status(&self, path: &Path) -> Result<ProjectStatus> {
        // Check for deprecation indicators
        let deprecated_indicators = ["DEPRECATED", "deprecated", ".deprecated"];
        for indicator in deprecated_indicators {
            if path.join(indicator).exists() {
                return Ok(ProjectStatus::Deprecated);
            }
        }

        // Check for archive indicators
        if path.join("ARCHIVED").exists() {
            return Ok(ProjectStatus::Archived);
        }

        // Check git activity (if available)
        if let Ok(Some(last_commit)) = self.get_last_commit(path).await {
            if let Ok(commit_time) = chrono::DateTime::parse_from_rfc3339(&last_commit) {
                let age = chrono::Utc::now().signed_duration_since(commit_time);

                if age.num_days() < 30 {
                    return Ok(ProjectStatus::Active);
                } else if age.num_days() < 180 {
                    return Ok(ProjectStatus::Maintenance);
                } else {
                    return Ok(ProjectStatus::Archived);
                }
            }
        }

        Ok(ProjectStatus::Unknown)
    }

    /// Get last commit timestamp
    async fn get_last_commit(&self, path: &Path) -> Result<Option<String>> {
        let git_dir = path.join(".git");
        if !git_dir.exists() {
            return Ok(None);
        }

        // Try to get last commit using git command
        let output = tokio::process::Command::new("git")
            .args(["-C", path.to_str().unwrap(), "log", "-1", "--format=%cI"])
            .output()
            .await;

        if let Ok(output) = output {
            if output.status.success() {
                let timestamp = String::from_utf8_lossy(&output.stdout).trim().to_string();
                if !timestamp.is_empty() {
                    return Ok(Some(timestamp));
                }
            }
        }

        Ok(None)
    }
}
